### PR TITLE
tlvf: Variable length lists support

### LINKED ISSUE #8: [task] Add TLVF support in multiple dynamic lists
TLVF currently only supports a single dynamic length array, and it has to be the last member in the class (be placed last on the buffer).
The reason for that is all the members are written directly on the buffer (and the class itself holds pointers to their location on the buffer), 
so allocating new element on a list that's not in the end of the buffer requires relocating all the data that comes after it.
This limitation is currently preventing us from properly implementing some types of messages, such as M1 & M2 frames, since those frames require multiple dynamic-length strings.

Currently, the workaround is defining fixed-size string in these messages. 
However, this workaround is only for the short term, since it doesn't match the specs, so it can only work with 

General idea of implementation:

1. for each dynamic size list in a class:
	1.1. listing (in python, not in generated code) all the class members that come after the list.
	1.2. In 'alloc_' method (for basic type or struct lists) and 'add_' method (for class type lists):
		1.2.1. Move the data and update pointer values of all members in 1.1 before the new elements' allocation on the buffer.
		1.2.2. memset the new allocated space to 0 (in order to avoid garbage in the new allocated elements).
		1.2.3. initializing the new elements and updating the relevant pointers and variable (existing code).

### PR DESCRIPTION
**Pull request by tomereli**
Screenshot from 2019-07-18 00-33-01
Fixes #8, #116

There are 2 types of variable (dynamic) length lists
in the TLV factory:
- known length
- unknown length

From the programmer point of view, these are almost identical. The main difference is that known length lists have the length variable part of the TLV.
Uknown length lists have to be the last member in the TLV so their length can be calculated
using the TLV length field.

TLVF currently only supports a single dynamic length list, and it has to be the last member in the class
(be placed last on the buffer). The reason for that is that all the members are written directly on the buffer (and the class itself holds pointers to their location on the buffer),
so allocating a new element on a list that's not in the end of the buffer requires relocating all the data that comes after it.

This limitation is currently preventing us from properly implementing some types of messages,
such as M1 & M2 frames, since those frames require multiple dynamic-length strings.

This PR adds full support for variable length lists as follows:
Dynamic length member have an alloc method which is used to allocate room on the buffer for the newly created list member.

This PR adds a `std::copy_n` to the alloc function which moves the remainder of the buffer
by the allocation size, and adds a marker which we use on every member which comes after this list to update its pointer.
For dynamic list members which are of class type, we also update the create method with a similar
`std::copy_n` since the create method adds the class on the buffer with the initial size, before
allocation method can be used.

Next, this PR aligns M1 and M2 TLVs to the WiFi Simple Setup (WSC) v2.0.6 specification as defined in the EasyMesh Multi-AP specification.

Finally, this PR adds README.md to the TLV factory with the design document and user guide.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

### CODE REVIEW

**Comment by tomereli** [comment]
Screenshot from 2019-07-18 00-33-01

File: framework/tlvf/tlvf.py [hunk 1]
```diff
@@ -445,12 +445,22 @@ def generateObject(self, obj_meta, dict_value):
                 if param_meta.error: self.abort(param_meta.error)
                 param_type = param_meta.type
                 param_type_info = param_meta.type_info
+                is_enum_type = False
                 if param_type and param_type.startswith("e"):
```
Comment (LiorAmram) [inline framework/tlvf/tlvf.py]: this line should be replaced with:
`if TypeInfo(param_type).type == TypeInfo.ENUM`
That provides you the functionality you implemented in the following lines.

**Review by LiorAmram — CHANGES_REQUESTED** [review]

**Review by morantr — CHANGES_REQUESTED** [review]
memmove() is banned function. according to: 
https://github.com/intel/safestringlib/wiki/SDL-List-of-Banned-Functions
Use instead std::move()
http://www.cplusplus.com/reference/algorithm/move/

**Review by morantr — DISMISSED** [review]
What about all the class members pointers that are allocated after the list?
I mean, they are set to a specific location on the buffer on the init function. And they are still pointing to the same place after you moved the memory. Changing them will still create a data override.

**Comment by tomereli** [comment]
> memmove() is banned function. according to:
> https://github.com/intel/safestringlib/wiki/SDL-List-of-Banned-Functions
> Use instead std::move()
> http://www.cplusplus.com/reference/algorithm/move/

C memmove is banned - C++ std::memmove is fine (basically all of the STL APIs pass SDL)

**Comment by tomereli** [comment]
> What about all the class members pointers that are allocated after the list?
> I mean, they are set to a specific location on the buffer on the init function. And they are still pointing to the same place after you moved the memory. Changing them will still create a data override.

They are handled in the alloc function - read the commit description - when we build the alloc function in the TLVF, we add the `std::memmove` and add a special comment, "marker", that is later used when we are adding the parameters that come after the list. For each such parameter we add code using the marker to update the pointer before the `std::memmove`.

I have a bug which @LiorAmram pointed out - when the list element type is a class, we don't do that - if this is what you refer to then nice catch! I'll fix it in a followup commit.

**Comment by morantr** [comment]
> > memmove() is banned function. according to:
> > https://github.com/intel/safestringlib/wiki/SDL-List-of-Banned-Functions
> > Use instead std::move()
> > http://www.cplusplus.com/reference/algorithm/move/
> 
> C memmove is banned - C++ std::memmove is fine (basically all of the STL APIs pass SDL)

According to here:
https://stackoverflow.com/questions/8380805/difference-between-string-h-and-cstring
it is said that:
"Table 55 describes the header <cstring>.
The contents are the same as the Standard C library header , with the change to memchr() specified in 21.7."

So basically it is the same C function with small change. I still think it is better to use std::move instead, it is more cpp way.

**Comment by morantr** [comment]
> > What about all the class members pointers that are allocated after the list?
> > I mean, they are set to a specific location on the buffer on the init function. And they are still pointing to the same place after you moved the memory. Changing them will still create a data override.
> 
> They are handled in the alloc function - read the commit description - when we build the alloc function in the TLVF, we add the `std::memmove` and add a special comment, "marker", that is later used when we are adding the parameters that come after the list. For each such parameter we add code using the marker to update the pointer before the `std::memmove`.
> 
> I have a bug which @LiorAmram pointed out - when the list element type is a class, we don't do that - if this is what you refer to then nice catch! I'll fix it in a followup commit.

Well, I looked at the generated code of list class type and didn't see that the member pointers are moved. So yes I meant that.


.

**Comment by LiorAmram** [comment]
This [issue](https://github.com/prplfoundation/prplMesh/issues/115) should be addressed in the future.
For now it's OK to keep it as a working standard.

File: framework/tlvf/tlvf.py [hunk 2]
```diff
@@ -726,6 +731,8 @@ def addClassParam(self, obj_meta, param_name, param_type, param_type_info, param
             if is_var_len:
                 # variable length list support - swap length value for calculating list size
                 length_type = obj_meta.children_types[param_length]
+                param_meta.list_index = obj_meta.list_index
```
Comment (LiorAmram) [inline framework/tlvf/tlvf.py]: This mechanism should also be added to dynamic lists (is_dynamic_len - unknown list length).

File: framework/tlvf/tlvf.py [hunk 3]
```diff
@@ -874,6 +881,12 @@ def addClassParam(self, obj_meta, param_name, param_type, param_type_info, param
                 if TypeInfo(param_type).type == TypeInfo.CLASS: #TODO:only if it's the last member of the class
                     lines_h.append( "std::shared_ptr<%s> create_%s();" % (param_type, param_name) ) #TODO: maybe change to 'create' and add '_entry' postfix
                     lines_cpp.append( "std::shared_ptr<%s> %s::create_%s() {" % (param_type, obj_meta.name, param_name) )
+                    lines_cpp.append( "%sif (m_%s__ > %s) {" %(self.getIndentation(1), self.MEMBER_LOCK_ORDER_COUNTER, param_meta.list_index) )
+                    lines_cpp.append( "%sTLVF_LOG(ERROR) << \"Out of order allocation for variable length list %s, abort!\";" %(self.getIndentation(2), param_name) )
+                    lines_cpp.append( "%sreturn nullptr;" %self.getIndentation(2))
+                    lines_cpp.append( "%s} else {" %self.getIndentation(1))
```
Comment (LiorAmram) [inline framework/tlvf/tlvf.py]: 'else' section is unnecessary, because of the return statement.

File: framework/tlvf/tlvf.py [hunk 4]
```diff
@@ -874,6 +881,12 @@ def addClassParam(self, obj_meta, param_name, param_type, param_type_info, param
...
+                    lines_cpp.append( "%sm_%s__ = %s;" %(self.getIndentation(2), self.MEMBER_LOCK_ORDER_COUNTER, param_meta.list_index) )
```
Comment (LiorAmram) [inline framework/tlvf/tlvf.py]: Might be better to first make all sanity tests (count == 0 and available space on buffer) and only then update the counter.

File: framework/tlvf/tlvf.py [hunk 5]
```diff
@@ -926,6 +939,12 @@ def addClassParam(self, obj_meta, param_name, param_type, param_type_info, param
                 else: #simple list
                     lines_h.append( "bool alloc_%s(size_t count = 1);" % (param_name) )
                     lines_cpp.append( "bool %s::alloc_%s(size_t count) {" % (obj_meta.name, param_name) )
+                    lines_cpp.append( "%sif (m_%s__ > %s) {;" %(self.getIndentation(1), self.MEMBER_LOCK_ORDER_COUNTER, param_meta.list_index) )
```
Comment (LiorAmram) [inline framework/tlvf/tlvf.py:942]: remove semicolon at the end of this line

File: framework/tlvf/tlvf.py [hunk 6]
```diff
@@ -926,6 +939,12 @@ def addClassParam(self, obj_meta, param_name, param_type, param_type_info, param
...
+                    lines_cpp.append( "%sTLVF_LOG(ERROR) << \"Out of order allocation for variable length list %s, abort!\";" %(self.getIndentation(2), param_name) )
+                    lines_cpp.append( "%sreturn false;" %self.getIndentation(2))
+                    lines_cpp.append( "%s} else {" %self.getIndentation(1))
```
Comment (LiorAmram) [inline framework/tlvf/tlvf.py]: same - remove else

File: framework/tlvf/tlvf.py [hunk 7]
```diff
@@ -926,6 +939,12 @@ def addClassParam(self, obj_meta, param_name, param_type, param_type_info, param
...
+                    lines_cpp.append( "%sm_%s__ = %s;" %(self.getIndentation(2), self.MEMBER_LOCK_ORDER_COUNTER, param_meta.list_index) )
```
Comment (LiorAmram) [inline framework/tlvf/tlvf.py]: same

**Review by LiorAmram — CHANGES_REQUESTED** [review]

**Comment by tomereli** [comment]
> memmove() is banned function. according to:
> https://github.com/intel/safestringlib/wiki/SDL-List-of-Banned-Functions
> Use instead std::move()
> http://www.cplusplus.com/reference/algorithm/move/

Done

**Comment by tomereli** [comment]
> > > What about all the class members pointers that are allocated after the list?
> > > I mean, they are set to a specific location on the buffer on the init function. And they are still pointing to the same place after you moved the memory. Changing them will still create a data override.
> > 
> > 
> > They are handled in the alloc function - read the commit description - when we build the alloc function in the TLVF, we add the `std::memmove` and add a special comment, "marker", that is later used when we are adding the parameters that come after the list. For each such parameter we add code using the marker to update the pointer before the `std::memmove`.
> > I have a bug which @LiorAmram pointed out - when the list element type is a class, we don't do that - if this is what you refer to then nice catch! I'll fix it in a followup commit.
> 
> Well, I looked at the generated code of list class type and didn't see that the member pointers are moved. So yes I meant that.
> 
> .

It all works fine now. Generated code does update the flag:
```
bool cPreferenceOperatingClasses::alloc_channel_list(size_t count) {
    if (m_lock_order_counter__ > 0) {;
        TLVF_LOG(ERROR) << "Out of order allocation for variable length list channel_list, abort!";
        return false;
    } else {
        m_lock_order_counter__ = 0;
    }
    if (count == 0) {
        TLVF_LOG(WARNING) << "can't allocate 0 bytes";
        return false;
    }
    size_t len = sizeof(uint8_t) * count;
    if(getBuffRemainingBytes() < len )  {
        TLVF_LOG(ERROR) << "Not enough available space on buffer - can't allocate";
        return false;
    }
    if (!m_parse__) {
        uint8_t *src = (uint8_t *)m_channel_list;
        uint8_t *dst = (uint8_t *)m_channel_list + len;
        size_t move_length = getBuffRemainingBytes(src) - len;
        std::copy_n(src, move_length, dst);
    }
    m_flags = (sFlags *)((uint8_t *)(m_flags) + len); // <----- update m_flags pointer
    m_channel_list_idx__ += count;
    *m_channel_list_length += count;
    m_buff_ptr__ += len;
    return true;
}
```

I also triggered channel selection from BML and changed the flags reason code to `OPERATION_WOULD_PREVENT_BACKHAUL_OPERATION_USING_SHARED_RADIO` to make sure it works, and it does:

image

**Review by LiorAmram — APPROVED** [review]

**Review by simonbaren — APPROVED** [review]

**Pull request merged**